### PR TITLE
Create PrimitiveMsg Factory

### LINKED ISSUE #1639: Create PrimitiveMsg Factory
### Description of the task
Create primitive_msg_factory[.h/.cpp] in software/primitive for creating `PrimitiveNewMsg`s. The file will contain functions like
`std::unique_ptr<PrimitiveMsg> createAutokickMovePrimitive(const Point &dest, const Angle &final_angle, double final_speed, double dribbler_speed, double kick_speed)` for all `*PrimitiveMsg` in primitive.proto except for `DirectControlPrimitiveMsg`

NOTE: while this will be called PrimitiveMsgFactory, it will create PrimitiveNewMsgs as defined in primitive.proto since as a step before integration for the new primitive protos. During integration, we will rename PrimitiveNewMsg to PrimitiveMsg

### Acceptance criteria

- [ ] New factory functions
- [ ] Tests

### Blocked By
#1620

### PR DESCRIPTION
**Pull request by kelvinkoon**
### Description

Create `primitive_msg_factory` to create messages for:
- `ChipPrimitiveMsg`
- `KickPrimitiveMsg`
- `MovePrimitiveMsg`
- `SpinningMovePrimitiveMsg`
- `AutochipMovePrimitiveMsg`
- `AutokickMovePrimitiveMsg`
- `StopPrimitiveMsg`

### Testing Done

Tests in `primitive_msg_factory_test.cpp`

### Resolved Issues

Resolves #1639 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/primitive/primitive_msg_factory.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    auto chip_primitive_msg = std::make_unique<ChipPrimitiveMsg>();
+
+    auto chip_origin_msg = std::make_unique<PointMsg>();
+    chip_origin_msg->set_x_meters(static_cast<float>(chip_origin.x()));
+    chip_origin_msg->set_y_meters(static_cast<float>(chip_origin.y()));
+    *(chip_primitive_msg->mutable_chip_origin()) = *chip_origin_msg;
+
+    auto chip_direction_msg = std::make_unique<AngleMsg>();
+    chip_direction_msg->set_radians(static_cast<float>(chip_direction.toRadians()));
+    *(chip_primitive_msg->mutable_chip_direction()) = *chip_direction_msg;
```
Comment (akhilveeraghanta) [inline src/software/primitive/primitive_msg_factory.cpp]: I think you can make use of some of the translation functions available in `software/proto/message_translation`
Specifically these ones found in `tbots_protobuf.h`
image

Ditto for all the other functions

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
Did a quick pass, overall looks good, just one comment on using message translation functions

File: src/software/primitive/primitive_msg_factory.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+#include "shared/proto/primitive.pb.h"
+#include "software/geom/angle.h"
+#include "software/geom/angular_velocity.h"
+#include "software/geom/point.h"
+#include "software/primitive/primitive.h"
+
+/**
+ * Enum for indicating the stop type for the Stop Primitive
+ */
+enum StopTypeMsg
```
Comment (jonathanlew) [inline src/software/primitive/primitive_msg_factory.h]: Msg is for protobuf messages
```suggestion
enum StopType
```

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Actually I think this should live in software/proto/primitive. Otherwise, good work!

File: src/software/primitive/primitive_msg_factory.h (hunk 2 above)
Comment (kelvinkoon) [reply src/software/primitive/primitive_msg_factory.h]: Moved to `proto/primitive` and fixed.

File: src/software/primitive/primitive_msg_factory.cpp (hunk 1 above)
Comment (kelvinkoon) [reply src/software/primitive/primitive_msg_factory.cpp]: Will save me some time in the future. Now using it.

File: src/software/proto/primitive/primitive_msg_factory.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+#include "software/primitive/primitive.h"
+
+/**
+ * Enum for indicating the stop type for the Stop Primitive
+ */
+enum StopType
+{
+    BRAKE,
+    COAST
+};
```
Comment (jonathanlew) [inline src/software/proto/primitive/primitive_msg_factory.h]: this should use MAKE_ENUM

File: src/software/proto/primitive/primitive_msg_factory.h [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ * Create a Chip Primitive Message
+ *
+ * @param chip_origin The location where the chip will be taken
+ * @param chip_direction The orientation the Robot will chip at
+ * @param chip_distance_meters The distance between the starting location
+ * of the chip and the location of the first bounce
+ *
+ * @return Pointer to Chip Primitive Message
+ */
+std::unique_ptr<ChipPrimitiveMsg> createChipPrimitiveMsg(const Point &chip_origin,
```
Comment (jonathanlew) [inline src/software/proto/primitive/primitive_msg_factory.h]: these should return PrimitiveNewMsg which is a `oneof` of ChipPrimitiveMsg, KickPrimitiveMsg, etc.

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/proto/primitive/primitive_msg_factory.h (hunk 3 above)
Comment (kelvinkoon) [reply src/software/proto/primitive/primitive_msg_factory.h]: Done.

File: src/software/proto/primitive/primitive_msg_factory.h (hunk 4 above)
Comment (kelvinkoon) [reply src/software/proto/primitive/primitive_msg_factory.h]: Done.

**Review by jonathanlew — APPROVED** [review]
LGTM! :+1:

**Review by akhilveeraghanta — APPROVED** [review]
🔥

**Pull request merged**